File: MB-Secretaria/providers_py/provider_retry.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, TypeVar

T = TypeVar("T")


@dataclass(frozen=True)
class ProviderRetryPolicy:
    max_attempts: int = 1
    base_delay_ms: int = 250


def with_provider_retry(operation: Callable[[int], T], policy: ProviderRetryPolicy = ProviderRetryPolicy()) -> tuple[T, int]:
    last_error: Exception | None = None
    for attempt in range(1, max(1, policy.max_attempts) + 1):
        try:
            return operation(attempt), attempt - 1
        except Exception as exc:
            last_error = exc
    assert last_error is not None
    raise last_error

File: MetaBrain/providers/provider-retry.ts
export type ProviderRetryPolicy = {
  max_attempts: number;
  base_delay_ms: number;
  retry_on_status: number[];
};

export const DEFAULT_PROVIDER_RETRY_POLICY: ProviderRetryPolicy = {
  max_attempts: 1,
  base_delay_ms: 250,
  retry_on_status: [408, 429, 500, 502, 503, 504],
};

export async function withProviderRetry<T>(
  operation: (attempt: number) => Promise<T>,
  policy: ProviderRetryPolicy = DEFAULT_PROVIDER_RETRY_POLICY,
): Promise<{ value: T; retry_count: number }> {
  let lastError: unknown;
  for (let attempt = 1; attempt <= Math.max(1, policy.max_attempts); attempt += 1) {
    try {
      const value = await operation(attempt);
      return { value, retry_count: attempt - 1 };
    } catch (error) {
      lastError = error;
      if (attempt >= policy.max_attempts) break;
      await new Promise((resolve) => setTimeout(resolve, policy.base_delay_ms * attempt));
    }
  }
  throw lastError;
}
